docs/dev/migration.md

```diff
@@ -25,9 +25,326 @@ under the License.
+return a `Collections.singletonList(MY_STATE)` in the `snaphshotState()`. The updated code for the `BufferingSink` 
+is included below:
+
+    public class BufferingSinkListCheckpointed implements 
+            SinkFunction<Tuple2<String, Integer>>,
+            ListCheckpointed<Tuple2<String, Integer>>, 
+            CheckpointedRestoring<ArrayList<Tuple2<String, Integer>>> {
+
+        private final int threashold;
+
+        private transient ListState<Tuple2<String, Integer>> checkpointedState;
+
+        private List<Tuple2<String, Integer>> bufferedElements;
+
+        public BufferingSinkListCheckpointed(int threashold) {
+            this.threashold = threashold;
+            this.bufferedElements = new ArrayList<>();
+        }
+
+        @Override
+        public void invoke(Tuple2<String, Integer> value) throws Exception {
+            this.bufferedElements.add(value);
+            if (bufferedElements.size() == threashold) {
+                for (Tuple2<String, Integer> element: bufferedElements) {
+                    // send it to the sink
+                }
+                bufferedElements.clear();
+            }
+        }
+
+        @Override
+        public List<Tuple2<String, Integer>> snapshotState(
+                long checkpointId, long timestamp) throws Exception {
+            return this.bufferedElements;
+        }
+
+        @Override
+        public void restoreState(List<Tuple2<String, Integer>> state) throws Exception {
+            if (!state.isEmpty()) {
+                this.bufferedElements.addAll(state);
+            }
+        }
+        
+        @Override
+        public void restoreState(ArrayList<Tuple2<String, Integer>> state) throws Exception {
+            // this is from the CheckpointedRestoring interface.
+            this.bufferedElements.addAll(state);
+        }
+    }
+
+As shown in the code, the updated function also implements the `CheckpointedRestoring` interface. This is for backwards 
+compatibility reasons and more details will be explained in the end of this section.
+
+##### CheckpointedFunction
+
+The `CheckpointedFunction` interface requires again the implementation of two methods:
+
+    void snapshotState(FunctionSnapshotContext context) throws Exception;
+
+    void initializeState(FunctionInitializationContext context) throws Exception;
+
+The difference here is that although, as in Flink-1.1, the `snapshotState()` is called when a checkpoint 
+is performed, the `initializeState()` which is the counterpart of the `restoreState()`, is called every time 
+the user-defined function is initialized and not only in the case that we are recovering from a failure. 
+Given this, the `initializeState()` is the place where the different types of the state can be initialized, 
+but also where the state recovery logic is included. The code for the `BufferingSink` implementing the 
+`CheckpointedFunction` interface is presented below.
```

Comment:
As in Flink-1.1, `snapshotState()` is called whenever a checkpoint 
is performed, but now `initializeState()` (which is the counterpart of the `restoreState()`) is called every time 
the user-defined function is initialized, rather than only in the case that we are recovering from a failure. 
Given this, `initializeState()` is not only the place where different types of state are initialized, 
but also where state recovery logic is included. An implementation of the `CheckpointedFunction` interface for `BufferingSink` is presented below.